x/ibc/04-channel/keeper/timeout.go

```diff
@@ -0,0 +1,175 @@
+		)
+	case types.UNORDERED:
+		ok = k.connectionKeeper.VerifyNonMembership(
+			ctx, connection, proofHeight, proof,
+			types.PacketAcknowledgementPath(packet.SourcePort(), packet.SourceChannel(), packet.Sequence()),
+		)
+	default:
+		panic("invalid channel ordering type")
+	}
+
+	if !ok {
+		return nil, errors.New("failed packet verification") // TODO: sdk.Error
+	}
+
+	k.deletePacketCommitment(ctx, packet.SourcePort(), packet.SourceChannel(), packet.Sequence())
+
+	if channel.Ordering == types.ORDERED {
+		channel.State = types.CLOSED
+		k.SetChannel(ctx, packet.SourcePort(), packet.SourceChannel(), channel)
+	}
+
+	return packet, nil
+}
+
+// TimeoutOnClose is called by a module in order to prove that the channel to
+// which an unreceived packet was addressed has been closed, so the packet will
+// never be received (even if the timeoutHeight has not yet been reached).
+func (k Keeper) TimeoutOnClose(
+	ctx sdk.Context,
+	packet exported.PacketI,
+	proofNonMembership,
+	proofClosed ics23.Proof,
+	proofHeight uint64,
+) (exported.PacketI, error) {
+	channel, found := k.GetChannel(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return nil, types.ErrChannelNotFound(k.codespace)
+	}
+
+	_, found = k.GetChannelCapability(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return nil, types.ErrChannelCapabilityNotFound(k.codespace)
+	}
+
+	// if !AuthenticateCapabilityKey(capabilityKey) {
+	//  return errors.New("invalid capability key") // TODO: sdk.Error
+	// }
+
+	if packet.DestChannel() != channel.Counterparty.ChannelID {
+		return nil, errors.New("invalid packet destination channel")
+	}
+
+	connection, found := k.connectionKeeper.GetConnection(ctx, channel.ConnectionHops[0])
+	if !found {
+		return nil, ics03types.ErrConnectionNotFound(k.codespace)
+	}
+
+	if packet.DestPort() != channel.Counterparty.PortID {
+		return nil, errors.New("invalid packet destination port")
+	}
+
+	if packet.DestPort() != channel.Counterparty.PortID {
+		return nil, errors.New("port id doesn't match with counterparty's")
+	}
+
+	commitment := k.GetPacketCommitment(ctx, packet.SourcePort(), packet.SourceChannel(), packet.Sequence())
+	if !bytes.Equal(commitment, packet.Data()) { // TODO: hash packet data
+		return nil, errors.New("packet hasn't been sent")
+	}
+
+	counterparty := types.NewCounterparty(packet.SourcePort(), packet.SourceChannel())
+	expectedChannel := types.NewChannel(
+		types.CLOSED, channel.Ordering, counterparty, channel.CounterpartyHops(), channel.Version,
+	)
+
+	bz, err := k.cdc.MarshalBinaryLengthPrefixed(expectedChannel)
+	if err != nil {
+		return nil, errors.New("failed to marshal expected channel")
+	}
+
+	if !k.connectionKeeper.VerifyMembership(
+		ctx, connection, proofHeight, proofClosed,
+		types.ChannelPath(channel.Counterparty.PortID, channel.Counterparty.ChannelID),
+		bz,
+	) {
+		return nil, types.ErrInvalidCounterpartyChannel(k.codespace)
+	}
+
+	if !k.connectionKeeper.VerifyNonMembership(
```

Comment:
This should do something different for ordered channels, this is a mistake in the spec

https://github.com/cosmos/ics/issues/287